cpp/src/arrow/compute/exec.cc

```diff
@@ -163,6 +163,34 @@ Result<ExecBatch> ExecBatch::Make(std::vector<Datum> values) {
       continue;
     }
 
+    if (length != value.length()) {
+      // all the arrays should have the same length
+      return -1;
+    }
```

Comment:
It's a recoverable error in the `ExecBatch::Make` loop when there are only scalar `Datum`s. I can change it to return an `optional<int64_t>` because I want to express that inference of the length failing is a non-fatal. `GroupBy` can use this, but ultimately it's `ExecBatch::Make` that performs the last validation and returns a more appropriate error `Status`.

Reply:
I added a commit that adds a reusable function for this.